### PR TITLE
[JENKINS-34246] Fix MultiBranchProjectFactory API to allow deletion or update of existing children

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-34246](https://issues.jenkins-ci.org/browse/JENKINS-34246)

@reviewbybees

### CODE REVIEW

File: src/main/java/jenkins/branch/MultiBranchProjectFactory.java [hunk 1]
```diff
@@ -46,21 +47,67 @@
...
+
+    /**
+     * Updates an existing project which matches {@link #recognizes}.
+     * @param project an existing project, perhaps created by this factory, perhaps not
+     * @param attributes a set of metadata attributes as added by {@link jenkins.scm.api.SCMSourceObserver.ProjectObserver#addAttribute}
+     * @param listener a way of reporting progress
+     */
+    public void updateExistingProject(MultiBranchProject<?,?> project, @Nonnull Map<String,Object> attributes, @Nonnull TaskListener listener) throws IOException, InterruptedException {}
+
+    @Deprecated
```
Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProjectFactory.java]: Not bothering with documentation, as the only implementation in production code was via `BySCMSourceCriteria` anyway.

File: src/main/java/jenkins/branch/OrganizationFolder.java [hunk 2]
```diff
@@ -158,41 +158,46 @@ public void addAttribute(String key, Object value) throws IllegalArgumentExcepti
                         }
                         @Override
                         public void complete() throws IllegalStateException, InterruptedException {
-                            MultiBranchProject<?,?> existing = observer.shouldUpdate(projectName);
-                            if (existing != null) {
-                                PersistedList<BranchSource> sourcesList = existing.getSourcesList();
-                                sourcesList.clear();
```
Comment (jglick) [inline src/main/java/jenkins/branch/OrganizationFolder.java]: [whitespace-ignoring diff](https://github.com/jenkinsci/branch-api-plugin/pull/45/files?w=1)

File: src/main/java/jenkins/branch/MultiBranchProjectFactory.java [hunk 3]
```diff
@@ -46,21 +47,67 @@
...
      * @param name a project name
      * @param scmSources a set of SCM sources as added by {@link jenkins.scm.api.SCMSourceObserver.ProjectObserver#addSource}
      * @param attributes a set of metadata attributes as added by {@link jenkins.scm.api.SCMSourceObserver.ProjectObserver#addAttribute}
      * @param listener a way of reporting progress
-     * @return a new uninitialized multibranch project (do not configure its {@link MultiBranchProject#getSourcesList} or call {@link MultiBranchProject#onCreatedFromScratch}), or null if unrecognized
+     * @return true if recognized
      */
+    public boolean recognizes(@Nonnull ItemGroup<?> parent, @Nonnull String name, @Nonnull List<? extends SCMSource> scmSources, @Nonnull Map<String,Object> attributes, @Nonnull TaskListener listener) throws IOException, InterruptedException {
+        if (Util.isOverridden(MultiBranchProjectFactory.class, getClass(), "createProject", ItemGroup.class, String.class, List.class, Map.class, TaskListener.class)) {
+            return createProject(parent, name, scmSources, attributes, listener) != null;
```
Comment (amuniz) [inline src/main/java/jenkins/branch/MultiBranchProjectFactory.java]: Where is `createProject` defined? The only one I see is this (but it does not return `boolean`):

```
public MultiBranchProject<?,?> createProject(@Nonnull ItemGroup<?> parent, @Nonnull String name, @Nonnull List<? extends SCMSource> scmSources, @Nonnull Map<String,Object> attributes, @Nonnull TaskListener listener) throws IOException, InterruptedException
```

I must be missing something.

Comment (jglick) [inline src/main/java/jenkins/branch/MultiBranchProjectFactory.java]: It is defined below, and deprecated. It returns `MultiBranchProject` or null, and here I am checking which.

Comment (amuniz) [inline src/main/java/jenkins/branch/MultiBranchProjectFactory.java]: Oooh, sorry, I missed the `!= null` at the end of the line :smile:

**Comment by amuniz** [comment]
:bee:

**Comment by abayer** [comment]
:bee:

**Pull request merged**